File: dotnet/utils/ImportTestXray.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace QAItest.Playwright.DotNet.Utils;

public static partial class ImportTestXray
{
    private static readonly HttpClient Http = new();

    [GeneratedRegex(@"[A-Z]{2,}-\d+")]
    private static partial Regex JiraKeyRegex();

    public static string SecretIdXray => Env("TEST_MGMT_SECRET_ID");
    public static string SecretIdJira => Env("ISSUE_TRACKER_SECRET_ID");
    public static string XrayBaseUrl => Env("TEST_MGMT_BASE_URL");
    public static string XrayGqlUrl => Env("TEST_MGMT_GRAPHQL_URL");
    public static string JiraBaseUrl => Env("ISSUE_TRACKER_BASE_URL");
    public static string FeaturesDir => FirstNonBlank(Env("FEATURES_DIR"), "tests/features");
    public static string LogsRoot => FirstNonBlank(Env("LOGS_ROOT"), "logs");

    public static void Log(string message) => Console.WriteLine(message);

    public static Dictionary<string, string> JiraAuthHeader()
    {
        var jiraCreds = string.IsNullOrWhiteSpace(SecretIdJira) ? new Dictionary<string, object>() : Aws.GetSecretDict(SecretIdJira);
        var email = FirstNonBlank(Env("ISSUE_TRACKER_EMAIL"),
            FirstNonBlank(Env("JIRA_EMAIL"),
                FirstNonBlank(jiraCreds.GetValueOrDefault("email")?.ToString(), jiraCreds.GetValueOrDefault("JIRA_EMAIL")?.ToString())));
        var token = FirstNonBlank(Env("ISSUE_TRACKER_TOKEN"),
            FirstNonBlank(Env("JIRA_TOKEN"),
                FirstNonBlank(jiraCreds.GetValueOrDefault("token")?.ToString(), jiraCreds.GetValueOrDefault("JIRA_TOKEN")?.ToString())));

        if (string.IsNullOrWhiteSpace(email) || string.IsNullOrWhiteSpace(token))
        {
            throw new InvalidOperationException("Missing issue tracker credentials (email/token).");
        }

        var encoded = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{email}:{token}"));
        return new Dictionary<string, string> { ["Authorization"] = $"Basic {encoded}" };
    }

    public static async Task<string> GetXrayTokenAsync()
    {
        if (string.IsNullOrWhiteSpace(XrayBaseUrl))
        {
            return string.Empty;
        }

        var xrayCreds = string.IsNullOrWhiteSpace(SecretIdXray) ? new Dictionary<string, object>() : Aws.GetSecretDict(SecretIdXray);
        var clientId = FirstNonBlank(Env("TEST_MGMT_CLIENT_ID"),
            FirstNonBlank(xrayCreds.GetValueOrDefault("client_id")?.ToString(), xrayCreds.GetValueOrDefault("XRAY_CLIENT_ID")?.ToString()));
        var clientSecret = FirstNonBlank(Env("TEST_MGMT_CLIENT_SECRET"),
            FirstNonBlank(xrayCreds.GetValueOrDefault("client_secret")?.ToString(), xrayCreds.GetValueOrDefault("XRAY_CLIENT_SECRET")?.ToString()));

        if (string.IsNullOrWhiteSpace(clientId) || string.IsNullOrWhiteSpace(clientSecret))
        {
            throw new InvalidOperationException("Missing test management credentials.");
        }

        var body = JsonSerializer.Serialize(new { client_id = clientId, client_secret = clientSecret });
        using var request = new HttpRequestMessage(HttpMethod.Post, $"{XrayBaseUrl.TrimEnd('/')}/authenticate")
        {
            Content = new StringContent(body, Encoding.UTF8, "application/json")
        };

        using var response = await Http.SendAsync(request);
        await EnsureSuccessAsync(response, "xray-auth");
        return (await response.Content.ReadAsStringAsync()).Trim().Trim('"');
    }

    public static async Task<Dictionary<string, object>> PostXrayGraphqlAsync(
        string token,
        string query,
        Dictionary<string, object>? variables = null,
        string operationName = "graphql")
    {
        if (string.IsNullOrWhiteSpace(XrayGqlUrl))
        {
            throw new InvalidOperationException("TEST_MGMT_GRAPHQL_URL is not configured.");
        }

        var body = JsonSerializer.Serialize(new { query, variables = variables ?? new Dictionary<string, object>() });
        using var request = new HttpRequestMessage(HttpMethod.Post, XrayGqlUrl)
        {
            Content = new StringContent(body, Encoding.UTF8, "application/json")
        };
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        using var response = await Http.SendAsync(request);
        await EnsureSuccessAsync(response, operationName);
        var payload = JsonSerializer.Deserialize<Dictionary<string, object>>(await response.Content.ReadAsStringAsync()) ?? new Dictionary<string, object>();
        if (payload.TryGetValue("errors", out var errors) && errors is not null && !string.IsNullOrWhiteSpace(errors.ToString()))
        {
            throw new InvalidOperationException($"[{operationName}] GraphQL errors: {errors}");
        }

        return payload;
    }

    public static async Task<List<Dictionary<string, object>>> JiraSearchIssuesAsync(string jql, List<string>? fields = null, int pageSize = 50)
    {
        if (string.IsNullOrWhiteSpace(JiraBaseUrl))
        {
            return [];
        }

        var fieldString = string.Join(",", fields is null || fields.Count == 0 ? ["key"] : fields);
        var uri = $"{JiraBaseUrl.TrimEnd('/')}/rest/api/3/search?jql={Uri.EscapeDataString(jql)}&maxResults={pageSize}&fields={Uri.EscapeDataString(fieldString)}";
        using var request = new HttpRequestMessage(HttpMethod.Get, uri);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        foreach (var header in JiraAuthHeader())
        {
            request.Headers.TryAddWithoutValidation(header.Key, header.Value);
        }

        using var response = await Http.SendAsync(request);
        await EnsureSuccessAsync(response, "jira-search");
        var payload = JsonSerializer.Deserialize<Dictionary<string, object>>(await response.Content.ReadAsStringAsync()) ?? new Dictionary<string, object>();
        return payload.TryGetValue("issues", out var issues) && issues is JsonElement element && element.ValueKind == JsonValueKind.Array
            ? element.EnumerateArray()
                .Where(item => item.ValueKind == JsonValueKind.Object)
                .Select(item => JsonSerializer.Deserialize<Dictionary<string, object>>(item.GetRawText()) ?? new Dictionary<string, object>())
                .ToList()
            : [];
    }

    public static string NormalizeLabel(string value)
    {
        return Regex.Replace(value ?? string.Empty, @"[^a-zA-Z0-9._-]+", "_").Trim('_').ToLowerInvariant();
    }

    public static List<Dictionary<string, object>> LoadCucumber(string path) => Xray.LoadCucumber(path);

    public static Dictionary<string, object> ExtractLabelsAndTestKeys(string cucumberFile)
    {
        var labels = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var jiraKeys = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var testKeys = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var statusCount = new Dictionary<string, int>
        {
            ["passed"] = 0,
            ["failed"] = 0,
            ["skipped"] = 0
        };

        if (!File.Exists(cucumberFile))
        {
            return new Dictionary<string, object>
            {
                ["labels"] = labels.ToList(),
                ["test_count"] = 0,
                ["jira_keys"] = new List<string>(),
                ["status_count"] = statusCount
            };
        }

        foreach (var feature in LoadCucumber(cucumberFile))
        {
            foreach (var element in AsObjectList(feature, "elements"))
            {
                var scenarioFailed = false;
                var scenarioPassed = true;
                foreach (var step in AsObjectList(element, "steps"))
                {
                    var status = AsNestedString(step, "result", "status").ToLowerInvariant();
                    if (status == "failed")
                    {
                        scenarioFailed = true;
                        scenarioPassed = false;
                    }
                    else if (status == "skipped")
                    {
                        scenarioPassed = false;
                    }
                }

                if (scenarioFailed)
                {
                    statusCount["failed"]++;
                }
                else if (scenarioPassed)
                {
                    statusCount["passed"]++;
                }
                else
                {
                    statusCount["skipped"]++;
                }

                foreach (var tag in AsObjectList(element, "tags"))
                {
                    var name = AsString(tag, "name").TrimStart('@').Trim();
                    if (string.IsNullOrWhiteSpace(name))
                    {
                        continue;
                    }

                    if (name.StartsWith("TEST_", StringComparison.Ordinal))
                    {
                        testKeys.Add(name["TEST_".Length..]);
                    }
                    else if (JiraKeyRegex().IsMatch(name))
                    {
                        jiraKeys.Add(name);
                    }
                    else
                    {
                        labels.Add(name);
                    }
                }
            }
        }

        return new Dictionary<string, object>
        {
            ["labels"] = labels.OrderBy(item => item).ToList(),
            ["test_count"] = testKeys.Count,
            ["jira_keys"] = jiraKeys.OrderBy(item => item).ToList(),
            ["status_count"] = statusCount
        };
    }

    public static List<string> ExtractLabelsFromReport(string cucumberPath)
    {
        return ExtractLabelsAndTestKeys(cucumberPath).TryGetValue("labels", out var labels) && labels is List<string> list ? list : [];
    }

    public static Dictionary<string, object> BuildDescriptionAdf(
        List<string> labels,
        int testCount,
        string environment,
        string dateString,
        string? apkVersion = null,
        string? apkBuildRuntime = null)
    {
        var lines = new List<string>
        {
            "Automated test execution summary.",
            string.Empty,
            $"Date: {dateString}",
            $"Environment: {environment}",
            $"Tests executed: {testCount}",
            $"Version: {(!string.IsNullOrWhiteSpace(apkVersion) ? apkVersion : "unknown")}"
        };

        if (!string.IsNullOrWhiteSpace(apkBuildRuntime))
        {
            lines.Add($"Build: {apkBuildRuntime}");
        }

        lines.Add(string.Empty);
        lines.Add($"Labels: {(labels.Count > 0 ? string.Join(", ", labels) : "none")}");
        var text = string.Join('\n', lines);

        return new Dictionary<string, object>
        {
            ["type"] = "doc",
            ["version"] = 1,
            ["content"] = new List<Dictionary<string, object>>
            {
                new()
                {
                    ["type"] = "paragraph",
                    ["content"] = new List<Dictionary<string, object>>
                    {
                        new() { ["type"] = "text", ["text"] = text }
                    }
                }
            }
        };
    }

    public static async Task Main(string[] args)
    {
        var report = Path.GetFullPath(args.Length > 0 ? args[0] : "reports/cucumber.json");
        if (!File.Exists(report))
        {
            Log($"Report not found: {report}");
            return;
        }

        var summary = ExtractLabelsAndTestKeys(report);
        var payload = new Dictionary<string, object>(summary)
        {
            ["report"] = report,
            ["features_dir"] = FeaturesDir,
            ["logs_root"] = LogsRoot,
            ["date"] = DateTime.UtcNow.ToString("yyyy-MM-dd"),
            ["failed_scenarios"] = Xray.SummarizeFailedScenarios(report)
        };

        var output = Path.Combine(Path.GetDirectoryName(report) ?? ".", $"{Path.GetFileNameWithoutExtension(report)}.summary.json");
        await File.WriteAllTextAsync(output, JsonSerializer.Serialize(payload, new JsonSerializerOptions { WriteIndented = true }));
        Log($"Summary written to {output}");
    }

    private static List<Dictionary<string, object>> AsObjectList(Dictionary<string, object> source, string key)
    {
        if (!source.TryGetValue(key, out var value) || value is not JsonElement element || element.ValueKind != JsonValueKind.Array)
        {
            return [];
        }

        return element.EnumerateArray()
            .Where(item => item.ValueKind == JsonValueKind.Object)
            .Select(item => JsonSerializer.Deserialize<Dictionary<string, object>>(item.GetRawText()) ?? new Dictionary<string, object>())
            .ToList();
    }

    private static string AsString(Dictionary<string, object> source, string key)
    {
        if (!source.TryGetValue(key, out var value) || value is null)
        {
            return string.Empty;
        }

        return value is JsonElement element
            ? element.ValueKind == JsonValueKind.String ? element.GetString() ?? string.Empty : element.ToString()
            : value.ToString() ?? string.Empty;
    }

    private static string AsNestedString(Dictionary<string, object> source, string nestedKey, string finalKey)
    {
        if (!source.TryGetValue(nestedKey, out var nested) || nested is not JsonElement element || element.ValueKind != JsonValueKind.Object)
        {
            return string.Empty;
        }

        var map = JsonSerializer.Deserialize<Dictionary<string, object>>(element.GetRawText()) ?? new Dictionary<string, object>();
        return AsString(map, finalKey);
    }

    private static string Env(string key) => Environment.GetEnvironmentVariable(key)?.Trim() ?? string.Empty;

    private static string FirstNonBlank(string? first, string? second)
    {
        return !string.IsNullOrWhiteSpace(first) ? first : second?.Trim() ?? string.Empty;
    }

    private static async Task EnsureSuccessAsync(HttpResponseMessage response, string operation)
    {
        if ((int)response.StatusCode >= 400)
        {
            throw new InvalidOperationException($"[{operation}] HTTP {(int)response.StatusCode}: {await response.Content.ReadAsStringAsync()}");
        }
    }
}

File: java/utils/ImportTestXray.java
package utils;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

public final class ImportTestXray {
  private static final ObjectMapper MAPPER = new ObjectMapper();
  private static final HttpClient HTTP = HttpClient.newHttpClient();
  private static final TypeReference<Map<String, Object>> MAP_TYPE = new TypeReference<>() {};
  private static final Pattern JIRA_KEY_RE = Pattern.compile("[A-Z]{2,}-\\d+");

  public static final String SECRET_ID_XRAY = env("TEST_MGMT_SECRET_ID");
  public static final String SECRET_ID_JIRA = env("ISSUE_TRACKER_SECRET_ID");
  public static final String XRAY_BASE_URL = env("TEST_MGMT_BASE_URL");
  public static final String XRAY_GQL_URL = env("TEST_MGMT_GRAPHQL_URL");
  public static final String JIRA_BASE_URL = env("ISSUE_TRACKER_BASE_URL");
  public static final String FEATURES_DIR = firstNonBlank(env("FEATURES_DIR"), "tests/features");
  public static final String LOGS_ROOT = firstNonBlank(env("LOGS_ROOT"), "logs");

  private ImportTestXray() {}

  public static void log(String message) {
    System.out.println(message);
  }

  public static Map<String, String> jiraAuthHeader() throws IOException {
    Map<String, Object> jiraCreds = SECRET_ID_JIRA.isBlank() ? Map.of() : Aws.getSecretDict(SECRET_ID_JIRA);
    String email = firstNonBlank(env("ISSUE_TRACKER_EMAIL"), firstNonBlank(env("JIRA_EMAIL"),
        firstNonBlank(asString(jiraCreds.get("email")), asString(jiraCreds.get("JIRA_EMAIL")))));
    String token = firstNonBlank(env("ISSUE_TRACKER_TOKEN"), firstNonBlank(env("JIRA_TOKEN"),
        firstNonBlank(asString(jiraCreds.get("token")), asString(jiraCreds.get("JIRA_TOKEN")))));
    if (email.isBlank() || token.isBlank()) {
      throw new IOException("Missing issue tracker credentials (email/token).");
    }
    String encoded = Base64.getEncoder().encodeToString((email + ":" + token).getBytes(StandardCharsets.UTF_8));
    return Map.of("Authorization", "Basic " + encoded);
  }

  public static String getXrayToken() throws IOException, InterruptedException {
    if (XRAY_BASE_URL.isBlank()) {
      return "";
    }
    Map<String, Object> xrayCreds = SECRET_ID_XRAY.isBlank() ? Map.of() : Aws.getSecretDict(SECRET_ID_XRAY);
    String clientId = firstNonBlank(env("TEST_MGMT_CLIENT_ID"),
        firstNonBlank(asString(xrayCreds.get("client_id")), asString(xrayCreds.get("XRAY_CLIENT_ID"))));
    String clientSecret = firstNonBlank(env("TEST_MGMT_CLIENT_SECRET"),
        firstNonBlank(asString(xrayCreds.get("client_secret")), asString(xrayCreds.get("XRAY_CLIENT_SECRET"))));
    if (clientId.isBlank() || clientSecret.isBlank()) {
      throw new IOException("Missing test management credentials.");
    }

    String body = MAPPER.writeValueAsString(Map.of("client_id", clientId, "client_secret", clientSecret));
    HttpRequest request = HttpRequest.newBuilder(URI.create(XRAY_BASE_URL.replaceAll("/+$", "") + "/authenticate"))
        .header("Content-Type", "application/json")
        .POST(HttpRequest.BodyPublishers.ofString(body))
        .build();
    HttpResponse<String> response = HTTP.send(request, HttpResponse.BodyHandlers.ofString());
    ensureSuccess(response, "xray-auth");
    return response.body().trim().replace("\"", "");
  }

  public static Map<String, Object> postXrayGraphql(String token, String query, Map<String, Object> variables, String operationName)
      throws IOException, InterruptedException {
    if (XRAY_GQL_URL.isBlank()) {
      throw new IOException("TEST_MGMT_GRAPHQL_URL is not configured.");
    }
    String body = MAPPER.writeValueAsString(Map.of("query", query, "variables", variables == null ? Map.of() : variables));
    HttpRequest request = HttpRequest.newBuilder(URI.create(XRAY_GQL_URL))
        .header("Authorization", "Bearer " + token)
        .header("Content-Type", "application/json")
        .POST(HttpRequest.BodyPublishers.ofString(body))
        .build();
    HttpResponse<String> response = HTTP.send(request, HttpResponse.BodyHandlers.ofString());
    ensureSuccess(response, operationName);
    Map<String, Object> payload = MAPPER.readValue(response.body(), MAP_TYPE);
    Object errors = payload.get("errors");
    if (errors instanceof List<?> list && !list.isEmpty()) {
      throw new IOException("[" + operationName + "] GraphQL errors: " + list);
    }
    return payload;
  }

  public static List<Map<String, Object>> jiraSearchIssues(String jql, List<String> fields, int pageSize)
      throws IOException, InterruptedException {
    if (JIRA_BASE_URL.isBlank()) {
      return List.of();
    }
    String fieldString = String.join(",", fields == null || fields.isEmpty() ? List.of("key") : fields);
    String url = JIRA_BASE_URL.replaceAll("/+$", "") + "/rest/api/3/search?jql=" + uriEncode(jql)
        + "&maxResults=" + pageSize + "&fields=" + uriEncode(fieldString);
    HttpRequest.Builder builder = HttpRequest.newBuilder(URI.create(url))
        .header("Accept", "application/json")
        .header("Content-Type", "application/json")
        .GET();
    for (Map.Entry<String, String> header : jiraAuthHeader().entrySet()) {
      builder.header(header.getKey(), header.getValue());
    }
    HttpResponse<String> response = HTTP.send(builder.build(), HttpResponse.BodyHandlers.ofString());
    ensureSuccess(response, "jira-search");
    Map<String, Object> payload = MAPPER.readValue(response.body(), MAP_TYPE);
    return asList(payload.get("issues"));
  }

  public static String normalizeLabel(String value) {
    return value == null ? "" : value.replaceAll("[^a-zA-Z0-9._-]+", "_").replaceAll("^_+|_+$", "").toLowerCase();
  }

  public static List<Map<String, Object>> loadCucumber(Path path) throws IOException {
    return Xray.loadCucumber(path);
  }

  public static Map<String, Object> extractLabelsAndTestKeys(Path cucumberFile) throws IOException {
    List<String> labels = new ArrayList<>();
    LinkedHashSet<String> jiraKeys = new LinkedHashSet<>();
    LinkedHashSet<String> testKeys = new LinkedHashSet<>();
    Map<String, Integer> statusCount = new LinkedHashMap<>();
    statusCount.put("passed", 0);
    statusCount.put("failed", 0);
    statusCount.put("skipped", 0);

    if (!Files.exists(cucumberFile)) {
      return Map.of("labels", labels, "test_count", 0, "jira_keys", List.of(), "status_count", statusCount);
    }

    for (Map<String, Object> feature : loadCucumber(cucumberFile)) {
      for (Map<String, Object> element : asList(feature.get("elements"))) {
        boolean scenarioFailed = false;
        boolean scenarioPassed = true;
        for (Map<String, Object> step : asList(element.get("steps"))) {
          String status = asString(asMap(step.get("result")).get("status")).toLowerCase();
          if ("failed".equals(status)) {
            scenarioFailed = true;
            scenarioPassed = false;
          } else if ("skipped".equals(status)) {
            scenarioPassed = false;
          }
        }
        if (scenarioFailed) {
          statusCount.put("failed", statusCount.get("failed") + 1);
        } else if (scenarioPassed) {
          statusCount.put("passed", statusCount.get("passed") + 1);
        } else {
          statusCount.put("skipped", statusCount.get("skipped") + 1);
        }

        for (Map<String, Object> tag : asList(element.get("tags"))) {
          String name = asString(tag.get("name")).replaceFirst("^@", "").trim();
          if (name.isBlank()) {
            continue;
          }
          if (name.startsWith("TEST_")) {
            testKeys.add(name.substring("TEST_".length()));
          } else if (JIRA_KEY_RE.matcher(name).matches()) {
            jiraKeys.add(name);
          } else {
            labels.add(name);
          }
        }
      }
    }

    Map<String, Object> payload = new LinkedHashMap<>();
    payload.put("labels", new ArrayList<>(new LinkedHashSet<>(labels)));
    payload.put("test_count", testKeys.size());
    payload.put("jira_keys", new ArrayList<>(jiraKeys));
    payload.put("status_count", statusCount);
    return payload;
  }

  public static List<String> extractLabelsFromReport(Path cucumberPath) throws IOException {
    Object labels = extractLabelsAndTestKeys(cucumberPath).get("labels");
    return labels instanceof List<?> list ? list.stream().map(String::valueOf).toList() : List.of();
  }

  public static Map<String, Object> buildDescriptionAdf(
      List<String> labels, int testCount, String envName, String dateStr, String apkVersion, String apkBuildRuntime) {
    List<String> lines = new ArrayList<>();
    lines.add("Automated test execution summary.");
    lines.add("");
    lines.add("Date: " + dateStr);
    lines.add("Environment: " + envName);
    lines.add("Tests executed: " + testCount);
    lines.add("Version: " + (apkVersion == null || apkVersion.isBlank() ? "unknown" : apkVersion));
    if (apkBuildRuntime != null && !apkBuildRuntime.isBlank()) {
      lines.add("Build: " + apkBuildRuntime);
    }
    lines.add("");
    lines.add("Labels: " + (labels == null || labels.isEmpty() ? "none" : String.join(", ", labels)));
    String text = String.join("\n", lines);

    return Map.of(
        "type", "doc",
        "version", 1,
        "content", List.of(Map.of(
            "type", "paragraph",
            "content", List.of(Map.of("type", "text", "text", text))
        ))
    );
  }

  public static void main(String[] args) throws IOException {
    Path report = Path.of(args.length > 0 ? args[0] : "reports/cucumber.json").toAbsolutePath();
    if (!Files.exists(report)) {
      log("Report not found: " + report);
      return;
    }

    Map<String, Object> summary = extractLabelsAndTestKeys(report);
    Map<String, List<String>> failed = Xray.summarizeFailedScenarios(report);
    Map<String, Object> payload = new LinkedHashMap<>();
    payload.put("report", report.toString());
    payload.put("features_dir", FEATURES_DIR);
    payload.put("logs_root", LOGS_ROOT);
    payload.put("date", LocalDate.now().toString());
    payload.putAll(summary);
    payload.put("failed_scenarios", failed);

    Path output = report.resolveSibling(report.getFileName().toString().replaceFirst("\\.json$", ".summary.json"));
    Files.writeString(output, MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(payload), StandardCharsets.UTF_8);
    log("Summary written to " + output);
  }

  private static void ensureSuccess(HttpResponse<String> response, String operation) throws IOException {
    if (response.statusCode() >= 400) {
      throw new IOException("[" + operation + "] HTTP " + response.statusCode() + ": " + response.body());
    }
  }

  @SuppressWarnings("unchecked")
  private static Map<String, Object> asMap(Object value) {
    return value instanceof Map<?, ?> map ? (Map<String, Object>) map : Map.of();
  }

  @SuppressWarnings("unchecked")
  private static List<Map<String, Object>> asList(Object value) {
    return value instanceof List<?> list ? (List<Map<String, Object>>) (List<?>) list : List.of();
  }

  private static String asString(Object value) {
    return value == null ? "" : String.valueOf(value);
  }

  private static String env(String key) {
    return System.getenv().getOrDefault(key, "").trim();
  }

  private static String firstNonBlank(String first, String second) {
    return first != null && !first.isBlank() ? first : (second == null ? "" : second);
  }

  private static String uriEncode(String value) {
    return URLEncoder.encode(value, StandardCharsets.UTF_8);
  }
}
